File: sdk/src/main/java/io/verloop/sdk/VerloopConfig.java
package io.verloop.sdk;

import android.content.SharedPreferences;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class VerloopConfig {
    private String userId;
    private String clientId;
    private String fcmToken;

    private String userName;
    private String userEmail;
    private String userPhone;

    private boolean isStaging = false;

    private ArrayList<CustomField> fields = new ArrayList<>();

    private String recipeId;

    private LiveChatButtonClickListener buttonOnClickListener;

    public enum Scope {
        USER,
        ROOM
    }

    public VerloopConfig(String clientId) {
        this.clientId = clientId;
    }

    public VerloopConfig(String clientId, String userId) {
        this.clientId = clientId;
        this.userId = userId;
    }

    String getUserId() {
        return userId;
    }

    String getClientId() {
        return clientId;
    }

    String getFcmToken() {
        return fcmToken;
    }

    boolean getStaging() {
        return isStaging;
    }

    LiveChatButtonClickListener getButtonOnClickListener() {
        return this.buttonOnClickListener;
    }

    void setUserId(String userId) {
        this.userId = userId;
    }

    public void setFcmToken(String fcmToken) {
        this.fcmToken = fcmToken;
    }

    public void setStaging(boolean staging) {
        this.isStaging = staging;
    }

    public void setUserEmail(String userEmail) {
        this.userEmail = userEmail;
    }

    public void setUserName(String userName) {
        this.userName = userName;
    }

    public void setUserPhone(String userPhone) {
        this.userPhone = userPhone;
    }

    /**
     * Call verloop.onStopChat() on destructor if button listener is added
     * @param buttonOnClickListener
     */
    public void setButtonOnClickListener(LiveChatButtonClickListener buttonOnClickListener) {
        this.buttonOnClickListener = buttonOnClickListener;
    }

    public void putCustomField(String key, String value, Scope scope) {
        fields.add(new CustomField(key, value, scope));
    }

    public void putCustomField(String key, String value) {
        fields.add(new CustomField(key, value, null));
    }

    public void setRecipeId(String id) {
        this.recipeId = id;
    }


    void save(SharedPreferences preferences) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(Verloop.CONFIG_CLIENT_ID, this.clientId);
        editor.putString(Verloop.CONFIG_USER_ID, this.userId);
        editor.putString(Verloop.CONFIG_FCM_TOKEN, this.fcmToken);
        editor.putString(Verloop.CONFIG_USER_NAME, this.userName);
        editor.putString(Verloop.CONFIG_USER_EMAIL, this.userEmail);
        editor.putString(Verloop.CONFIG_USER_PHONE, this.userPhone);
        editor.putBoolean(Verloop.CONFIG_STAGING, this.isStaging);
        editor.putString(Verloop.CONFIG_RECIPE_ID, this.recipeId);


        JSONObject object = new JSONObject();
        for (CustomField field : fields) {
            try {
                if (field.scope != null) {
                    JSONObject innerObject = new JSONObject();
                    JSONObject scopeObject = new JSONObject();

                    scopeObject.put("scope", field.scope.name().toLowerCase());

                    innerObject.put("value", field.value);
                    innerObject.put("options", scopeObject);

                    object.put(field.key, innerObject);
                } else {
                    object.put(field.key, field.value);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        editor.putString(Verloop.CONFIG_FIELDS, object.toString());
        editor.apply();
    }

    class CustomField {
        String key, value;
        Scope scope;

        public CustomField(String key, String value, Scope scope) {
            this.key = key;
            this.value = value;
            this.scope = scope;
        }
    }
}

File: sdk/src/main/java/io/verloop/sdk/VerloopConfig.kt
package io.verloop.sdk

import android.os.Parcelable
import io.verloop.sdk.model.HeaderConfig
import io.verloop.sdk.model.LogLevel
import kotlinx.parcelize.Parcelize
import java.util.*

@Parcelize
data class VerloopConfig private constructor(
    var clientId: String?,
    var userId: String?,
    var fcmToken: String?,
    var userName: String?,
    var userEmail: String?,
    var userPhone: String?,
    var recipeId: String?,
    var department: String?,
    var isStaging: Boolean = false,
    var closeExistingChat: Boolean = false,
    var overrideUrlClick: Boolean = false,
    var overrideHeaderLayout: Boolean = false,
    var openMenuWidgetOnStart: Boolean = false,
    var headerConfig: HeaderConfig? = null,
    var fields: ArrayList<CustomField> = ArrayList(),
    var allowFileDownload: Boolean = false
) : Parcelable {

    var logLevel: LogLevel = LogLevel.WARNING

    var buttonOnClickListener: LiveChatButtonClickListener? = null
    var chatUrlClickListener: LiveChatUrlClickListener? = null

    var logEventListener: LiveLogEventListener? = null
        private set

    @Deprecated("Use builder instead")
    constructor(clientId: String) : this(
        clientId,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        false,
        false,
        false,
        false,
        false,
        null,
        ArrayList(),
        false
    )

    @Deprecated("Use builder instead")
    constructor(clientId: String, userId: String) : this(clientId) {
        this.userId = userId
    }

    fun putCustomField(key: String, value: String, scope: Scope) {
        fields.add(CustomField(key, value, scope))
    }

    fun putCustomField(key: String, value: String) {
        fields.add(CustomField(key, value, null))
    }

    /**
     * Callback for button click event from within the chat
     * @param buttonOnClickListener
     */
    fun setButtonClickListener(buttonOnClickListener: LiveChatButtonClickListener) {
        this.buttonOnClickListener = buttonOnClickListener
    }

    /**
     * Callback for url click event from within the chat
     * @param urlClickListener
     */
    fun setUrlClickListener(
        urlClickListener: LiveChatUrlClickListener,
        overrideUrlClick: Boolean = false
    ) {
        this.chatUrlClickListener = urlClickListener
        this.overrideUrlClick = overrideUrlClick
    }

    /**
     * Callback for log events from within the chat
     * @param logEventListener
     */
    fun setLogEventListener(
        logEventListener: LiveLogEventListener,
        logLevel: LogLevel
    ) {
        this.logEventListener = logEventListener
        this.logLevel = logLevel
    }

    enum class Scope {
        USER, ROOM
    }

    @Parcelize
    class CustomField(var key: String?, var value: String?, var scope: Scope?) : Parcelable {

        constructor() : this(
            null, null, Scope.USER
        )

    }

    data class Builder(
        var clientId: String? = null,
        var userId: String? = null,
        var fcmToken: String? = null,
        var userName: String? = null,
        var userEmail: String? = null,
        var userPhone: String? = null,
        var recipeId: String? = null,
        var department: String? = null,
        var isStaging: Boolean = false,
        var closeExistingChat: Boolean = false,
        var overrideUrlClick: Boolean = false,
        var overrideHeaderLayout: Boolean = false,
        var openMenuWidgetOnStart: Boolean = false,
        var headerConfig: HeaderConfig? = null,
        var fields: ArrayList<CustomField> = ArrayList(),
        var allowFileDownload: Boolean = false,
    ) {
        fun clientId(clientId: String?) = apply { this.clientId = clientId }
        fun userId(userId: String?) = apply {
            if (!userId.equals(""))
                this.userId = userId
        }

        fun fcmToken(fcmToken: String?) = apply { this.fcmToken = fcmToken }
        fun userName(userName: String?) = apply { this.userName = userName }
        fun userEmail(userEmail: String?) = apply { this.userEmail = userEmail }
        fun userPhone(userPhone: String?) = apply { this.userPhone = userPhone }
        fun recipeId(recipeId: String?) = apply { this.recipeId = recipeId }
        fun department(department: String?) = apply { this.department = department }

        @Deprecated("Not required anymore", ReplaceWith(""))
        fun isStaging(isStaging: Boolean) = apply { this.isStaging = isStaging }

        fun closeExistingChat(closeExistingChat: Boolean) =
            apply { this.closeExistingChat = closeExistingChat }

        fun overrideUrlClick(overrideUrlClick: Boolean) =
            apply { this.overrideUrlClick = overrideUrlClick }

        /**
         * Override the toolbar configuration received from server with toolbar_verloop.xml defined in main app.
         * Can't use along with headerConfig. Either use headerConfig or overrideHeaderLayout
         * @param overrideHeaderLayout
         */
        fun overrideHeaderLayout(overrideHeaderLayout: Boolean) =
            apply { this.overrideHeaderLayout = overrideHeaderLayout }

        fun allowFileDownload(allowFileDownload: Boolean) =
            apply { this.allowFileDownload = allowFileDownload }

        fun openMenuWidgetOnStart(openMenuWidgetOnStart: Boolean) =
            apply { this.openMenuWidgetOnStart = openMenuWidgetOnStart }

        /**
         * Override the toolbar configuration received from server with provided header configuration.
         * Can't use along with overrideHeaderLayout. Either use overrideHeaderLayout or headerConfig
         * @param headerConfig
         */
        fun headerConfig(headerConfig: HeaderConfig?) = apply { this.headerConfig = headerConfig }

        fun fields(fields: ArrayList<CustomField>) = apply { this.fields = fields }

        @Throws(VerloopException::class)
        fun build(): VerloopConfig {
            if (clientId.isNullOrEmpty()) throw VerloopException("Client id cannot be null or empty")
            return VerloopConfig(
                clientId,
                userId,
                fcmToken,
                userName,
                userEmail,
                userPhone,
                recipeId,
                department,
                isStaging,
                closeExistingChat,
                overrideUrlClick,
                overrideHeaderLayout,
                openMenuWidgetOnStart,
                headerConfig,
                fields,
                allowFileDownload
            )
        }
    }
}